### PR TITLE
[SMSP] Adding new SMSP problems

### PR DESCRIPTION
**Pull request by AminosDz**
This is a first draft of implementing general purpose methods for all SMS problems:

- Instances are expressed using the smsp_instance decorator
- Algorithms rely on the compute_objective methods that hides the objective from the resolution algorithm
- Algorithm also use the objective_insert which returns the objective value when inserting a job in a certain position regardless of the objective

PS: the methods call compute_objective too much since it relies on the cache (wiCi_index and wiTi_index). Better performances should be achieved when implementing smsp_solution decorator which creates performant methods to compute the objective.

### CODE REVIEW

**Comment by AminosDz** [comment]
@tahaarbaoui @MouradTerzi 

I finished refactoring the simulation operators (12 operators, 3 per objective). I also added all the generic problems for SMSP, this includes the new objective wiFi.

Next steps for me are the following:

- Reduce the number of operators using a generic function from 12 operator to 6 objectives, where the operator compute_objective_remove_insert and swap will cover any objective (First approach will include the ifs).
- For performance reasons, we can use decorators to define only the needed objective like we've done with single_instance, only this time we will use exec() to define exactly and only the needed function.
- Refactoring method names, setting up the conventions and reporting them + finishing documentation.

Once these features are implemented I'll tag you in order to merge this with develop since PMSP (at least) will depend on this work. Since the functionality won't change, if you approve it at this point, I'll merge it once done.

**Comment by AminosDz** [comment]
@tahaarbaoui @MouradTerzi 

It is all done for this pull request for me. If you approve it I'll merge it with the develop branch so that the others can use what I've implemented.

File: src/pyscheduling/SMSP/SingleMachine.py [hunk 1]
```diff
@@ -440,485 +448,119 @@ def toJSON(self):
...
-
-        return wiCi
-
-    def total_weighted_lateness(self, instance: SingleInstance, startIndex: int = 0):
-        """Fills the job_schedule with the correct sequence of start_time and completion_time of each job and returns the total weighted lateness
+        objective = instance.get_objective()
+        obj_cache = self.objectives_map.get(objective, -1)
+        if obj_cache == -1: # No cache is used
+            return startIndex, None
+        if obj_cache is None:  # Iniates wiCi_index to the size of job_schedule
```
Comment (MouradTerzi) [inline src/pyscheduling/SMSP/SingleMachine.py:486]: I think, for the comment message in green, we have to replace wiCi_index by obj_cache !

Comment (MouradTerzi) [reply src/pyscheduling/SMSP/SingleMachine.py:486]: I think, for the comment message in green, we have to replace wiCi_index by obj_cache !

File: src/pyscheduling/SMSP/SingleMachine.py [hunk 2]
```diff
@@ -440,485 +448,119 @@ def toJSON(self):
...
-            if hasattr(instance, 'S'):
-                setupTime = instance.S[job_prev_i][job_i]
-            else:
-                setupTime = 0
-            proc_time = instance.P[job_i]
-            ci = startTime + setupTime + proc_time
-            wiTi += instance.W[job_i]*max(ci-instance.D[job_i], 0)
-
-            job_prev_i = job_i
+            ci, obj, job_prev_i,  = 0, 0, self.job_schedule[startIndex].id
```
Comment (MouradTerzi) [inline src/pyscheduling/SMSP/SingleMachine.py:509]: When we assign the value of self.job_schedule[startIndex].id to job_prev_i, this means that the job_schedule is not empty (contains one element), why in this case ci and obj are equal to 0 ?

File: src/pyscheduling/SMSP/SingleMachine.py (hunk 1 above)
Comment (AminosDz) [reply src/pyscheduling/SMSP/SingleMachine.py:486]: Yes I'll change it

File: src/pyscheduling/SMSP/SingleMachine.py (hunk 2 above)
Comment (AminosDz) [reply src/pyscheduling/SMSP/SingleMachine.py:509]: In this case, startIndex = 0. Therefore the job_prev_i is the first job in the schedule. 
This is to allow computing the first setup S[i][i]. And the previous ci and obj are just starting so they're equal to 0 (there's no job before startIndex).

**Pull request merged**